influxdata/tokenizer.go

```diff
@@ -7,91 +7,187 @@ import (
+	// next to be read. Data in buf[len(buf):cap(buf)] is
+	// available for reading into.
+	r1 int
+
+	// escBuf holds a buffer for unescaped characters.
+	escBuf []byte
 
 	// err holds any non-EOF error that was returned from rd.
 	err error
 }
 
-// take returns the next slice of bytes within t.buf that are in the given set
-// starting at t.r + from, reading more data as needed.
-//
-// It does not update t.r.
-func (t *Tokenizer) take(set *byteSet, from int) []byte {
-	n := from
+// NewTokenizer returns a tokenizer that splits the line-protocol text
+// inside buf. The text in buf represent a complete line (or multiple lines) - that
+// is, the tokenization will not work correctly if presented with a line
+// split at an arbitrary place.
+func NewTokenizerWithBytes(buf []byte) *Tokenizer {
+	return &Tokenizer{
+		buf:      buf,
+		complete: true,
+		escBuf:   make([]byte, 0, 512),
+	}
+}
+
+func NewTokenizer(r io.Reader) *Tokenizer {
+	return &Tokenizer{
+		rd: r,
+	}
+}
+
+// take returns the next slice of bytes that are in the given set
+// reading more data as needed. It updates t.r1.
+func (t *Tokenizer) take(set *byteSet) []byte {
+	// Note: use a relative index for start because absolute
+	// indexes aren't stable (the contents of the buffer can be
+	// moved when reading more data).
+	start := t.r1 - t.r0
 outer:
 	for {
-		if !t.ensure(n + 1) {
+		if !t.ensure(1) {
 			break
 		}
-		buf := t.buf[t.r+n:]
+		buf := t.buf[t.r1:]
 		for i, c := range buf {
 			if !set.get(c) {
-				n += i
+				t.r1 += i
 				break outer
 			}
 		}
-		n += len(buf)
+		t.r1 += len(buf)
 	}
-	return t.buf[t.r+from : t.r+n]
+	return t.buf[t.r0+start : t.r1]
+}
+
+// takeEsc is like take except that escaped characters also count as
+// part of the set. The escapeTable determines which characters
+// can be escaped.
+// It returns the unescaped string (unless t.skipping is true, in which
+// case it doesn't need to go to the trouble of unescaping it).
+func (t *Tokenizer) takeEsc(set *byteSet, escapeTable *[256]byte) []byte {
+	// start holds the offset from r0 of the start of the taken slice.
+	// Note that we can't use t.r1 directly, because the offsets can change
+	// when the buffer is grown.
+	start := t.r1 - t.r0
+
+	// startUnesc holds the offset from t0 of the start of the most recent
+	// unescaped segment.
+	startUnesc := start
+
+	// startEsc holds the index into r.escBuf of the start of the escape buffer.
+	startEsc := len(t.escBuf)
+outer:
+	for {
+		if !t.ensure(1) {
+			break
+		}
+		buf := t.buf[t.r1:]
+		for i := 0; i < len(buf); i++ {
+			c := buf[i]
+			if c != '\\' {
+				if !set.get(c) {
+					// We've found the end, so we're done here.
+					t.r1 += i
+					break outer
+				}
+				continue
+			}
+			if i+1 >= len(buf) {
+				// Not enough room in the buffer. Try reading more so that
+				// we can see the next character.
+				if !t.ensure(i + 2) {
+					// No character to escape, so leave the \ intact.
+					t.r1 = len(t.buf)
```

Comment:
`+=` no?

`t.r1 += len(t.buf)`

Reply:
No, I believe `=` is right here (note that it's `t.buf` here, not `buf`. When `t.ensure` returns false, it means that we've got to the end of the input and `t.buf` contains all the characters that it's ever going to. What this line is doing is consuming all of the rest of the input (up to and including the last `\` character). An equivalent way of writing this lin, I think, would be:

     t.r1 += i+1

but in my head, making `r1` equal the end of the buffer was a bit clearer. I'll try to clarify the comment here.

Reply:
Ah. It's `t.buf` - I was reading it as `len(buf)` as you guessed. 

But, my spidey sense is still firing. Tell me where I'm getting this wrong. Some` off by one` confusion?

```
// if i+1 >= len(buf) == true then ...
i == len(buf) - 1
t.buf[len(t.buf) - 1]  == buf[i] == '\'
// let's say there's one more byte to read for the source of t.buf
// next then we call ensure() in the if
if !t.ensure(i + 2) == FALSE // because requested 2 more bytes but only 1
// now
t.buf[len(t.buf) - 2] == buf[i] == '\'
// so there's one character at the end of t.buf now that we could read and escape.
```

The last byte isn't _not read_ but it seems it would be escaped still when it should be unescaped by this method.

Reply:
> if !t.ensure(i + 2) == FALSE // because requested 2 more bytes but only 1

I think this is where your reasoning isn't quite there. Note that the number we're passing to `t.ensure` is a count, not an index. So, we already know that `t.ensure(i+1)` is guaranteed to succeed because we know there are at least `i+1` bytes in the buffer. So `t.ensure(i+2)` is asking for exactly _one_ byte beyond the end of the buffer, not two (after all, we just want to peek at the single extra character after the `\` that we already know is there). So if there's at least one more byte in the source, then `t.ensure(i+2)` will succeed, so all's good. If there aren't any more bytes available, then `t.ensure(i+2)` will fail, and all is still good.

Am I missing something?

Reply:
nope - you're right. I had the off by one mistake.

Reply:
I added a note to the comment above to try to make this a bit more obvious.